File: MiniDB_Projects/Team_XO/internal/planner/optimizer.go
package planner

import (
	"fmt"
	"math"

	"minidb/internal/sql"
	"minidb/internal/types"
)

// Optimizer builds physical plans using table statistics from a Provider.
type Optimizer struct {
	provider Provider
}

// NewOptimizer constructs an optimizer over the given table provider.
func NewOptimizer(p Provider) *Optimizer { return &Optimizer{provider: p} }

type relation struct {
	name  string
	alias string
	info  TableInfo
}

func (r relation) matches(ref sql.ColumnRef) bool {
	if ref.Table != "" {
		return ref.Table == r.alias || ref.Table == r.name
	}
	return r.info.Schema().ColumnIndex(ref.Name) >= 0
}

func aliasOr(name, alias string) string {
	if alias != "" {
		return alias
	}
	return name
}

// BuildSelect produces a physical plan for sel, choosing scan methods, join
// algorithm and join order by estimated cost.
func (o *Optimizer) BuildSelect(sel *sql.Select) (Plan, error) {
	info0, ok := o.provider.TableInfo(sel.From)
	if !ok {
		return nil, fmt.Errorf("planner: unknown table %q", sel.From)
	}
	rel0 := relation{name: sel.From, alias: aliasOr(sel.From, sel.FromAlias), info: info0}

	conjuncts := flatten(sel.Where)

	var plan Plan
	if sel.Join == nil {
		plan = o.buildScan(rel0, assignConjuncts(conjuncts, []relation{rel0})[0])
	} else {
		info1, ok := o.provider.TableInfo(sel.Join.Table)
		if !ok {
			return nil, fmt.Errorf("planner: unknown table %q", sel.Join.Table)
		}
		rel1 := relation{name: sel.Join.Table, alias: aliasOr(sel.Join.Table, sel.Join.Alias), info: info1}
		buckets := assignConjuncts(conjuncts, []relation{rel0, rel1})
		left := o.buildScan(rel0, buckets[0])
		right := o.buildScan(rel1, buckets[1])
		plan = o.buildJoin(left, right, rel0, rel1, sel.Join)
	}

	return o.topLevel(sel, plan), nil
}

// topLevel wraps the scan/join plan in an aggregate or projection as the SELECT
// list dictates.
func (o *Optimizer) topLevel(sel *sql.Select, child Plan) Plan {
	if len(sel.Items) == 1 && sel.Items[0].IsCount {
		return &CountAgg{Child: child}
	}
	if len(sel.Items) == 1 && sel.Items[0].Star {
		return child
	}
	return &Projection{Child: child, Items: sel.Items}
}

// buildScan chooses between a sequential scan and an index scan for one relation
// given the predicates that reference only that relation.
func (o *Optimizer) buildScan(rel relation, conjuncts []sql.Expr) Plan {
	rows := float64(rel.info.RowCount())
	if rows < 1 {
		rows = 1
	}

	sel := 1.0
	for _, c := range conjuncts {
		sel *= selectivity(c, rel)
	}

	seq := &SeqScan{
		Table:  rel.name,
		Alias:  rel.alias,
		Filter: foldAnd(conjuncts),
		rows:   rows * sel,
		cost:   rows, // a sequential scan must read every tuple
	}

	// Look for an equality predicate on an indexed column to consider an index
	// scan instead.
	for i, c := range conjuncts {
		col, key, ok := equalityOnColumn(c, rel)
		if !ok || !rel.info.HasIndex(col) {
			continue
		}
		eqSel := selectivity(c, rel)
		residual := foldAnd(removeAt(conjuncts, i))
		idxRows := rows * sel
		idxCost := rows*eqSel + 1 // touch only matching tuples via the B+Tree
		if idxCost < seq.cost {
			return &IndexScan{
				Table:    rel.name,
				Alias:    rel.alias,
				Col:      col,
				ColName:  rel.info.Schema().Columns[col].Name,
				Key:      key,
				Residual: residual,
				rows:     idxRows,
				cost:     idxCost,
			}
		}
	}
	return seq
}

// buildJoin estimates output size and selects a join algorithm and orientation.
func (o *Optimizer) buildJoin(left, right Plan, lRel, rRel relation, j *sql.JoinClause) Plan {
	lRef, rRef := orientKeys(j, lRel, rRel)
	lCol := lRel.info.Schema().ColumnIndex(lRef.Name)
	rCol := rRel.info.Schema().ColumnIndex(rRef.Name)

	lRows := left.EstRows()
	rRows := right.EstRows()
	dl := math.Max(float64(lRel.info.DistinctValues(lCol)), 1)
	dr := math.Max(float64(rRel.info.DistinctValues(rCol)), 1)
	joinSel := 1.0 / math.Max(dl, dr)
	outRows := lRows * rRows * joinSel

	hashCost := lRows + rRows + outRows

	// Index nested-loop join is attractive when the inner side is indexed on the
	// join key, turning the inner scan into a point lookup.
	bestNLJCost := math.Inf(1)
	var nljOuter, nljInner Plan
	var nljOuterRef, nljInnerRef sql.ColumnRef
	if rRel.info.HasIndex(rCol) {
		if c := lRows + lRows; c < bestNLJCost {
			bestNLJCost = c
			nljOuter, nljInner = left, right
			nljOuterRef, nljInnerRef = lRef, rRef
		}
	}
	if lRel.info.HasIndex(lCol) {
		if c := rRows + rRows; c < bestNLJCost {
			bestNLJCost = c
			nljOuter, nljInner = right, left
			nljOuterRef, nljInnerRef = rRef, lRef
		}
	}

	if bestNLJCost < hashCost {
		return &NestedLoopJoin{
			Left: nljOuter, Right: nljInner,
			LeftKey: nljOuterRef, RightKey: nljInnerRef,
			rows: outRows, cost: bestNLJCost,
		}
	}

	// Hash join: build the hash table on the smaller input.
	bl, br := left, right
	blRef, brRef := lRef, rRef
	if rRows > lRows {
		bl, br = right, left
		blRef, brRef = rRef, lRef
	}
	return &HashJoin{
		Left: bl, Right: br,
		LeftKey: blRef, RightKey: brRef,
		rows: outRows, cost: hashCost,
	}
}

// orientKeys returns the join condition refs ordered as (left-relation ref,
// right-relation ref).
func orientKeys(j *sql.JoinClause, lRel, rRel relation) (sql.ColumnRef, sql.ColumnRef) {
	if lRel.matches(j.Left) {
		return j.Left, j.Right
	}
	return j.Right, j.Left
}

// selectivity estimates the fraction of rows a predicate keeps, using distinct
// counts for equality and fixed heuristics otherwise.
func selectivity(e sql.Expr, rel relation) float64 {
	be, ok := e.(sql.BinaryExpr)
	if !ok {
		return 1.0
	}
	switch be.Op {
	case "=":
		if col, _, ok := equalityOnColumn(be, rel); ok {
			d := rel.info.DistinctValues(col)
			if d < 1 {
				d = 1
			}
			return 1.0 / float64(d)
		}
		return 0.1
	case "!=":
		return 0.9
	case "<", "<=", ">", ">=":
		return 0.33
	default:
		return 1.0
	}
}

// equalityOnColumn recognises predicates of the form `column = literal` (in
// either operand order) on a column of rel, returning its index and the key.
func equalityOnColumn(e sql.Expr, rel relation) (int, types.Value, bool) {
	be, ok := e.(sql.BinaryExpr)
	if !ok || be.Op != "=" {
		return 0, types.Value{}, false
	}
	if ref, ok := be.Left.(sql.ColumnRef); ok {
		if lit, ok := be.Right.(sql.Literal); ok && rel.matches(ref) {
			return rel.info.Schema().ColumnIndex(ref.Name), lit.Value, true
		}
	}
	if ref, ok := be.Right.(sql.ColumnRef); ok {
		if lit, ok := be.Left.(sql.Literal); ok && rel.matches(ref) {
			return rel.info.Schema().ColumnIndex(ref.Name), lit.Value, true
		}
	}
	return 0, types.Value{}, false
}

// flatten splits an AND-chain into its conjuncts.
func flatten(e sql.Expr) []sql.Expr {
	if e == nil {
		return nil
	}
	be, ok := e.(sql.BinaryExpr)
	if !ok || be.Op != "AND" {
		return []sql.Expr{e}
	}
	return append(flatten(be.Left), flatten(be.Right)...)
}

// foldAnd recombines conjuncts into a single expression (nil if empty).
func foldAnd(cs []sql.Expr) sql.Expr {
	if len(cs) == 0 {
		return nil
	}
	out := cs[0]
	for _, c := range cs[1:] {
		out = sql.BinaryExpr{Op: "AND", Left: out, Right: c}
	}
	return out
}

func removeAt(cs []sql.Expr, i int) []sql.Expr {
	out := make([]sql.Expr, 0, len(cs)-1)
	out = append(out, cs[:i]...)
	return append(out, cs[i+1:]...)
}

// assignConjuncts routes each conjunct to the single relation it references.
// Cross-relation predicates (which our grammar expresses through JOIN ... ON)
// are dropped from per-table buckets.
func assignConjuncts(conjuncts []sql.Expr, rels []relation) [][]sql.Expr {
	buckets := make([][]sql.Expr, len(rels))
	for _, c := range conjuncts {
		refs := columnRefs(c)
		matched := -1
		single := true
		for _, ref := range refs {
			for ri, rel := range rels {
				if rel.matches(ref) {
					if matched == -1 {
						matched = ri
					} else if matched != ri {
						single = false
					}
				}
			}
		}
		if single && matched >= 0 {
			buckets[matched] = append(buckets[matched], c)
		}
	}
	return buckets
}

func columnRefs(e sql.Expr) []sql.ColumnRef {
	switch v := e.(type) {
	case sql.ColumnRef:
		return []sql.ColumnRef{v}
	case sql.BinaryExpr:
		return append(columnRefs(v.Left), columnRefs(v.Right)...)
	default:
		return nil
	}
}

File: MiniDB_Projects/Team_Cash/src/optimizer.cpp
#include "optimizer.h"

#include <stdexcept>

namespace minidb {

std::unique_ptr<PlanNode> Optimizer::scanFor(const std::string& table,
                                             const std::vector<Condition>& where,
                                             const Condition** consumed) {
    *consumed = nullptr;
    TableInfo* info = catalog_.get(table);
    if (!info) throw std::runtime_error("unknown table '" + table + "'");
    const std::string& pk = info->schema.columns[0].name;

    for (const Condition& c : where) {
        if (c.op == "=" && c.hasLiteral) {
            bool onPk = c.left.name == pk && (c.left.table.empty() || c.left.table == table);
            if (onPk && info->isIndexed()) {
                auto node = std::make_unique<PlanNode>();
                node->kind = PlanKind::IndexScan;
                node->table = table;
                node->key = c.literal;
                node->estRows = 1;
                *consumed = &c;
                return node;
            }
        }
    }
    auto node = std::make_unique<PlanNode>();
    node->kind = PlanKind::SeqScan;
    node->table = table;
    node->estRows = info->rowCount;
    return node;
}

std::unique_ptr<PlanNode> Optimizer::optimize(const SelectStmt& stmt) {
    std::unique_ptr<PlanNode> plan;

    if (!stmt.hasJoin) {
        const Condition* consumed = nullptr;
        plan = scanFor(stmt.table, stmt.where, &consumed);
        std::vector<Condition> residual;
        for (const Condition& c : stmt.where)
            if (&c != consumed) residual.push_back(c);
        if (!residual.empty()) {
            auto f = std::make_unique<PlanNode>();
            f->kind = PlanKind::Filter;
            f->conditions = residual;
            f->child = std::move(plan);
            plan = std::move(f);
        }
    } else {
        TableInfo* leftInfo = catalog_.get(stmt.table);
        TableInfo* rightInfo = catalog_.get(stmt.joinTable);
        if (!leftInfo || !rightInfo) throw std::runtime_error("unknown table in join");

        const Condition* c1 = nullptr;
        const Condition* c2 = nullptr;
        auto leftBase = scanFor(stmt.table, stmt.where, &c1);
        auto rightBase = scanFor(stmt.joinTable, stmt.where, &c2);

        auto join = std::make_unique<PlanNode>();
        join->kind = PlanKind::Join;
        join->joinCond = stmt.joinCond;
        // Smaller table becomes the inner (buffered) relation.
        if (rightInfo->rowCount <= leftInfo->rowCount) {
            join->left = std::move(leftBase);
            join->right = std::move(rightBase);
        } else {
            join->left = std::move(rightBase);
            join->right = std::move(leftBase);
        }
        plan = std::move(join);

        std::vector<Condition> residual;
        for (const Condition& c : stmt.where)
            if (&c != c1 && &c != c2) residual.push_back(c);
        if (!residual.empty()) {
            auto f = std::make_unique<PlanNode>();
            f->kind = PlanKind::Filter;
            f->conditions = residual;
            f->child = std::move(plan);
            plan = std::move(f);
        }
    }

    if (!stmt.columns.empty()) {
        auto p = std::make_unique<PlanNode>();
        p->kind = PlanKind::Project;
        p->columns = stmt.columns;
        p->child = std::move(plan);
        plan = std::move(p);
    }
    return plan;
}

std::string explainPlan(const PlanNode* node, int depth) {
    std::string pad(depth * 2, ' ');
    switch (node->kind) {
        case PlanKind::SeqScan:
            return pad + "SeqScan(" + node->table + ")  [reads all ~" +
                   std::to_string(node->estRows) + " rows]";
        case PlanKind::IndexScan:
            return pad + "IndexScan(" + node->table + ", key=" + node->key.toString() +
                   ")  [B+ tree lookup, O(log n)]";
        case PlanKind::Filter:
            return pad + "Filter(" + std::to_string(node->conditions.size()) + " cond)\n" +
                   explainPlan(node->child.get(), depth + 1);
        case PlanKind::Project: {
            std::string cols;
            for (size_t i = 0; i < node->columns.size(); ++i)
                cols += (i ? ", " : "") + node->columns[i].name;
            return pad + "Project(" + cols + ")\n" + explainPlan(node->child.get(), depth + 1);
        }
        case PlanKind::Join:
            return pad + "NestedLoopJoin\n" + explainPlan(node->left.get(), depth + 1) + "\n" +
                   explainPlan(node->right.get(), depth + 1);
    }
    return pad + "?";
}

}  // namespace minidb
